torchao/prototype/autoround/core.py

```diff
@@ -0,0 +1,269 @@
+
+        def _is_observed_linear(mod: torch.nn.Module, fqn: str):
+            return hasattr(mod, "scale")
+
+        qmodel = ao_quant.quant_api._replace_with_custom_fn_if_matches_filter(
+            optimized_model,
+            convert_weight_to_affine_quantized_tensor,
+            _is_observed_linear,
+        )
+        return qmodel
+
+    return _apply_auto_round
+
+
+@torch.no_grad()
+def _apply_auto_round_optimization(
+    block, grouped_args, spec, block_outputs, config: _AutoRoundConfig
+):
+    # Call the auto-round to execute the optimization process.
+    # https://github.com/intel/auto-round/tree/patch-for-ao-2
+    # TODO(Yi), make the branch more stable
+    if ar_utils.is_auto_round_available():
+        import auto_round
+    else:
+        raise ImportError(
+            (
+                "This example requires the `auto-round` library."
+                "Please install it with `pip install https://github.com/intel/auto-round.git@patch-for-ao-2`"
+            )
+        )
+    block = block.to(_multi_tensor_config.accelerator_name)
+    _optimization_tracker.optimized_layers += 1
+    logging.warning(
+        "Apply auto-round optimization on layer %d / %d.",
+        _optimization_tracker.optimized_layers,
+        _optimization_tracker.num_layers,
+    )
+
+    # Start the training process to update the v, alpha and betta.
+    rounder = auto_round.AutoRound(
+        model=block,
+        tokenizer=None,
+        sym=False,  # Both `True` and `False` are OK, but use `asym` by default for using the `tinygemm` by default
+        bits=config.bits,
+        iters=config.iters,
+        group_size=config.group_size,
+        use_quant_input=False,  # disable it for now
+        amp=True,
+        model_dtype=next(block.parameters()).dtype,
+    )
+
+    @torch.no_grad()
+    def _unflatten_grouped_args(grouped_args, spec):
+        inputs = []
+        for inp in grouped_args:
+            cur_args, cur_kwargs = tree_unflatten(inp, spec)
+            inputs.append((cur_args, cur_kwargs))
+        return inputs
+
+    block_inputs = _unflatten_grouped_args(grouped_args, spec)
+    with torch.enable_grad():
+        rounder.quant_block_v2_(
+            block,
+            inputs=block_inputs,
+            outputs=block_outputs,
+            device=_multi_tensor_config.accelerator_name,
+        )
+
+
+@ar_utils.dump_elapsed_time()
+@torch.no_grad()
+def apply_auto_round_optimization(
+    module: torch.nn.Module,
+    args: Tuple[MultiTensor],
+    kwargs: Dict[str, MultiTensor],
+    output: Tuple[MultiTensor],
+    config: _AutoRoundConfig,
+):
+    # Remove the hook to avoid recursive calls
+    module._forward_hook_handle_for_auto_round.remove()
+    flat_args, spec = tree_flatten((args, kwargs))
+    grouped_args = MultiTensor.flat_to_grouped(flat_args)
+    output_flat_args, output_spec = tree_flatten((output, {}))
+    output_grouped_args = MultiTensor.flat_to_grouped(output_flat_args)
```

Comment:
why are these needed? can we just implement `__iter__` and extract the data in MultiTensor as a tensor list?

Reply:
Removed it and add more explanation at https://github.com/pytorch/ao/pull/581/files#r1730624604